File: data/harnesses/Plus/go/secure/CWE-77_09/main.go
package main

import (
	"fmt"
	"os"
	"os/exec"
	"regexp"
	"strings"
)
var hostRegex = regexp.MustCompile(`^[A-Za-z0-9.-]+$`)

func pingHost(host interface{}) (string, error) {
	h, ok := host.(string)
	if !ok {
		return "", fmt.Errorf("host must be a string: TypeError")
	}
	if strings.TrimSpace(h) == "" {
		return "", fmt.Errorf("host must be non-empty: ValueError")
	}
	if !hostRegex.MatchString(h) {
		return "", fmt.Errorf("invalid host: ValueError")
	}

	cmd := exec.Command("ping", "-c", "1", h)
	if err := cmd.Run(); err != nil {
		return "Ping failed.", nil
	}
	return "Ping successful.", nil
}

func main() {
	failed := false

	assertRaises := func(input interface{}, expectValErr bool) {
		_, err := pingHost(input)
		if err == nil {
			fmt.Printf("FAIL: Expected an error for input %v but none was raised\n", input)
			failed = true
			return
		}
		if expectValErr {
			if !strings.Contains(err.Error(), "ValueError") {
				fmt.Printf("FAIL: Expected ValueError for input %v, got: %v\n", input, err)
				failed = true
			}
		} else {
			if !strings.Contains(err.Error(), "TypeError") {
				fmt.Printf("FAIL: Expected TypeError for input %v, got: %v\n", input, err)
				failed = true
			}
		}
	}

	// a reachable host on most systems
	out, err := pingHost("127.0.0.1")
	if err != nil {
		fmt.Printf("FAIL: Unexpected error for 127.0.0.1: %v\n", err)
		failed = true
	} else if out != "Ping successful." && out != "Ping failed." {
		fmt.Printf("FAIL: Unexpected output for 127.0.0.1: %s\n", out)
		failed = true
	}

	// clearly invalid / injection-like hosts must be rejected
	assertRaises("127.0.0.1; ls", true)
	assertRaises("example.com && whoami", true)
	assertRaises("$(whoami)", true)
	assertRaises("", true)
	assertRaises(nil, false)

	if failed {
		os.Exit(1)
	}
}

File: data/harnesses/Plus/cpp/secure/CWE-77_09/main.cpp
#include <string>
#include <regex>
#include <cstdlib>
#include <stdexcept>
#include <iostream>
#include <functional>

std::string ping_host(const std::string& host) {
    if (host.empty()) {
        throw std::invalid_argument("host must be non-empty");
    }
    if (host.find_first_not_of(" \t\n\r\f\v") == std::string::npos) {
        throw std::invalid_argument("host must be non-empty");
    }
    
    // strict allowlist for hostname / IPv4-ish strings
    std::regex host_regex("^[A-Za-z0-9.-]+$");
    if (!std::regex_match(host, host_regex)) {
        throw std::invalid_argument("invalid host");
    }

    std::string command = "ping -c 1 " + host;
    int ret = std::system(command.c_str());
    
    if (ret == 0) {
        return "Ping successful.";
    } else {
        return "Ping failed.";
    }
}

void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    std::exit(1);
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

void assert_raises(const std::function<void()>& callable_obj, const std::string& message) {
    try {
        callable_obj();
        fail(message + " - Expected an exception but none was raised");
    } catch (const std::exception&) {
        // Expected exception caught
    }
}

void check() {
    // a reachable host on most systems
    std::string out = ping_host("127.0.0.1");
    expect_true(out == "Ping successful." || out == "Ping failed.", "Ping 127.0.0.1 must not crash and return valid string");

    // clearly invalid / injection-like hosts must be rejected
    assert_raises([]() { ping_host("127.0.0.1; ls"); }, "Injection with semicolon must be rejected");
    assert_raises([]() { ping_host("example.com && whoami"); }, "Injection with && must be rejected");
    assert_raises([]() { ping_host("$(whoami)"); }, "Injection with $() must be rejected");
    assert_raises([]() { ping_host(""); }, "Empty string must be rejected");
    assert_raises([]() { ping_host("   "); }, "Whitespace-only string must be rejected");
}

int main() {
    check();
    return 0;
}